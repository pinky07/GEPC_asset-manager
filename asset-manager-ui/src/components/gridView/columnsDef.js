import NodeNameCellRenderer from './nodeNameCellRenderer';
import MixHeaderRenderer from './mixHeaderRenderer';
import ValueHeaderRenderer from './valueHeaderRenderer';

const sortAscending = '<i class="fa fa-sort-asc fa-3" aria-hidden="true"/>';
const sortDescending = '<i class="fa fa-sort-desc fa-3" aria-hidden="true"/>';
const sortUnSort = '<i class="fa fa-refresh fa-3" aria-hidden="true"/>';

export const percentageFormatter = params => {
  return isNaN(params.value)
    ? ''
    : Number(params.value / 100).toLocaleString(undefined, {
        style: 'percent',
        minimumFractionDigits: 0,
      });
};

export const columns = [
  {
    field: 'accountgroupname',
    headerName: 'Node Name',
    width: 350,
    cellRendererFramework: NodeNameCellRenderer,
    sortingOrder: [null, null],
    icons: {
      sortUnSort,
    },
    pinned: 'left',
  },
  {
    field: 'aa_category',
    headerName: 'Asset Category',
    width: 184,
    sortingOrder: ['asc', 'desc'],
    icons: {
      sortAscending,
      sortDescending,
    },
    pinned: 'left',
  },
  {
    field: 'accountgroupname',
    headerName: 'Asset Class Alias',
    width: 350,
    sortingOrder: ['asc', 'desc'],
    icons: {
      sortAscending,
      sortDescending,
    },
    pinned: 'left',
  },
  {
    field: 'aa_model_benchmark',
    headerName: 'AAMB',
    width: 200,
    sortingOrder: ['asc', 'desc'],
    icons: {
      sortAscending,
      sortDescending,
    },
    pinned: 'left',
  },
  {
    field: 'policy_value',
    headerName: 'Policy',
    width: 200,
    sortingOrder: ['asc', 'desc'],
    icons: {
      sortAscending,
      sortDescending,
    },
    pinned: 'left',
    valueFormatter: percentageFormatter,
  },
  {
    field: 'actual_mv',
    headerName: 'Value',
    width: 200,
    headerComponentFramework: ValueHeaderRenderer,
    sortingOrder: ['asc', 'desc'],
    icons: {
      sortAscending,
      sortDescending,
    },
    pinned: 'left',
    valueFormatter: percentageFormatter,
  },
];

export const mixColumn = {
  field: 'mix',
  width: 200,
  editable: true,
  headerName: 'Mix',
  headerComponentFramework: MixHeaderRenderer,
  sortingOrder: ['asc', 'desc'],
  icons: {
    sortAscending,
    sortDescending,
  },
  valueFormatter: percentageFormatter,
};
