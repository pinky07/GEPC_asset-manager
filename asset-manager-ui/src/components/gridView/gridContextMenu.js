import React from 'react';
import { connect } from 'react-redux';
import { ContextMenu, Item } from 'react-contexify';
import PropTypes from 'prop-types';

import { addNodeFromGrid } from '../../actions';

export class GridContextMenu extends React.Component {
  render() {
    return (
      <ContextMenu id="grid_menu_id">
        <Item onClick={this.props.addNodeFromGrid}>Add a Node</Item>
      </ContextMenu>
    );
  }
}

GridContextMenu.propTypes = {
  addNodeFromGrid: PropTypes.func.isRequired,
};

export default connect(null, { addNodeFromGrid })(GridContextMenu);
