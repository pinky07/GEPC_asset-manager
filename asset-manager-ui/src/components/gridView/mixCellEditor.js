import React from 'react';
//import { connect } from 'react-redux';
//import PropTypes from 'prop-types';

export default class MixCellEditor extends React.Component {
  static defaultProps = {
    value: ''
  };

  constructor(props) {
    super(props);

    this.cancelBeforeStart = this.props.charPress && ('1234567890'.indexOf(this.props.charPress) < 0);

    let value = this.props.value;
    if (!this.cancelBeforeStart && this.props.charPress) {
      value = value + this.props.charPress;
    }
    this.state = {
      value
    };
  }

  componentDidMount() {
    this.refs.input.addEventListener('keydown', this.onKeyDown);
    this.focus();
  }

  componentDidUpdate() {
    this.focus();
  }

  componentWillUnmount() {
    this.refs.input.removeEventListener('keydown', this.onKeyDown);
  }

  focus() {
    setTimeout(() => {
      this.refs.input.focus();
      this.refs.input.setSelectionRange(this.state.value.length, this.state.value.length);
    })
  }

  getValue() {
    return this.state.value;
  }

  isCancelBeforeStart() {
    return this.cancelBeforeStart;
  }

  onKeyDown = event => {
    if (this.isLeftOrRight(event)) {
      event.stopPropagation();
      return;
    }

    if (!this.isKeyPressedNumeric(event) && event.keyCode !== 8) {
      if (event.preventDefault) event.preventDefault();
    }
  }

  isLeftOrRight(event) {
    return [37, 39].indexOf(event.keyCode) > -1;
  }

  onChange = event => {
    this.setState({value: event.target.value});
  }

  getCharCodeFromEvent(event) {
    event = event || window.event;
    return (typeof event.which === "undefined") ? event.keyCode : event.which;
  }

  isCharNumeric(charStr) {
    return !!/\d/.test(charStr);
  }

  isKeyPressedNumeric(event) {
    const charCode = this.getCharCodeFromEvent(event);
    const charStr = event.key ? event.key : String.fromCharCode(charCode);
    return this.isCharNumeric(charStr);
  }

  render() {
    return (
      <input ref="input"
             value={this.state.value}
             onChange={this.onChange}
             onKeyDown={this.onKeyDown}
             style={{width: "100%"}}
      />
    );
  }
}
