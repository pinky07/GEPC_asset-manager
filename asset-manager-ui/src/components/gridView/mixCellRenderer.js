import React from 'react';
import PropTypes from 'prop-types';

class MixCellRenderer extends React.Component {
  render() {
    return (
      <div>
        {Number(this.props.value.mix1.value/100).toLocaleString(undefined,{style: 'percent', minimumFractionDigits:0})}
      </div>
    );
  }
}

MixCellRenderer.propTypes = {
  value: PropTypes.object,
};

export default MixCellRenderer;