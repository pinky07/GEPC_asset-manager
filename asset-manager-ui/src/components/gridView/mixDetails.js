import {
  Card,
  CardBlock,
  CardTitle,
  Input,
  InputGroup,
  InputGroupAddon,
} from 'reactstrap';
import React from 'react';

class MixDetails extends React.Component {
  render() {
    return (
      <div className="nodeDetails mixDetails">
        <Card>
          <CardBlock>
            <CardTitle>Mix Details</CardTitle>
            <InputGroup>
              <InputGroupAddon>Comments</InputGroupAddon>
              <Input />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>Mix Name</InputGroupAddon>
              <Input />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>Effective Date</InputGroupAddon>
              <Input />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>Mix Type</InputGroupAddon>
              <Input />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>Market Value</InputGroupAddon>
              <Input />
            </InputGroup>
          </CardBlock>
        </Card>
      </div>
    );
  }
}

export default MixDetails;
