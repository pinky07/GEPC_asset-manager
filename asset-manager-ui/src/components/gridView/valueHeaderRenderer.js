import React from 'react';
import { connect } from 'react-redux';
import { ContextMenuProvider } from 'react-contexify';
import PropTypes from 'prop-types';

export class ValueHeaderRenderer extends React.Component {
  constructor(props) {
    super(props);
    this.props.column.addEventListener('sortChanged', this.onSortChanged);
    this.state = {
      sorted: '',
    };
  }

  componentWillUnmount() {
    this.props.column.removeEventListener('sortChanged', this.onSortChanged);
  }

  onClickHeader = event => {
    if (event.target.textContent !== '...') {
      this.props.progressSort(event.shiftKey);
    }
  };

  onSortChanged = () => {
    const sorted = this.props.column.getSort();
    this.setState({
      sorted,
    });
  };

  sortAscButton = event => {
    this.props.setSort('asc', event.shiftKey);
  };

  sortDescButton = event => {
    this.props.setSort('desc', event.shiftKey);
  };

  renderMenu() {
    if (this.props.gridData.length === 0) {
      return null;
    }
    return (
      <ContextMenuProvider id="value_menu_id" renderTag="span" event="onClick">
        <button>...</button>
      </ContextMenuProvider>
    );
  }

  renderSortElements = () => {
    let sortElements = [];
    if (this.props.enableSorting) {
      let downArrowClass = `ag-header-icon ag-sort-descending-icon ${this.state
        .sorted === 'desc'
        ? ' active'
        : ' ag-hidden'}`;
      let upArrowClass = `ag-header-icon ag-sort-ascending-icon ${this.state
        .sorted === 'asc'
        ? ' active'
        : ' ag-hidden'}`;

      sortElements.push(
        <span
          key={`down${this.props.displayName}`}
          className={upArrowClass}
          onClick={this.sortAscButton}>
          <span className="ag-icon ag-icon-asc" />
        </span>
      );
      sortElements.push(
        <span
          key={`up${this.props.displayName}`}
          className={downArrowClass}
          onClick={this.sortDescButton}>
          <span className="ag-icon ag-icon-desc" />
        </span>
      );
    }
    return sortElements;
  };

  render() {
    let desc =
      this.state.sorted === 'desc' ? ' ag-header-cell-sorted-desc' : '';
    let asc = this.state.sorted === 'asc' ? ' ag-header-cell-sorted-asc' : '';
    let none = this.state.sorted === '' ? ' ag-header-cell-sorted-none' : '';

    return (
      <div className={`ag-cell-label-container ${desc} ${asc} ${none}`}>
        <div className="ag-header-cell-label" onClick={this.onClickHeader}>
          <span className="ag-header-cell-text">
            {this.props.column.colDef.headerName}
          </span>
          {this.renderSortElements()}
          {this.renderMenu()}
        </div>
      </div>
    );
  }
}

ValueHeaderRenderer.propTypes = {
  column: PropTypes.object.isRequired,
  gridData: PropTypes.array.isRequired,
  enableSorting: PropTypes.bool,
  progressSort: PropTypes.func,
  setSort: PropTypes.func,
  displayName: PropTypes.string,
};

const mapStateToProps = state => {
  return {
    ...state.allocationGrid,
  };
};

export default connect(mapStateToProps, null)(ValueHeaderRenderer);
