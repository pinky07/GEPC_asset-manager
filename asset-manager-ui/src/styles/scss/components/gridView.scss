.gridView {
  .dropdownPlanAnalysis {
    margin-left: 350px;
    width: 184px;
    display: inline-block;
  }

  .gridContainer {
    padding-top: .5em;
    width: 100%;
    height: 800px;
    padding-bottom: 2em;
    overflow: hidden;
  }

  .ag-header-row {
    color: white;
    font-weight: bold;
    background-color: $blue-1 !important;
  }

  .ag-header-cell-label {
    align-items: center;
    justify-content: center;
    height: 100%;
  }

  .ag-row.ag-row-even {
    background-color: $green-1;
  }

  .ag-row.ag-row-odd {
    background-color: $green-2;
  }

  .ag-react-container {
    .nav-link {
      padding: 0;
    }
  }

  i {
    display: inline-block;
    color: white;
    margin-left: 0.5em;
  }
}
