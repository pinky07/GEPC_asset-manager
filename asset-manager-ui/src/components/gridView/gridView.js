import { Col, Row } from 'reactstrap';
import { AgGridReact } from 'ag-grid-react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import React from 'react';
import Select from 'react-select';
import _ from 'lodash';

import NodeNameHeaderMenu from './nodeNameHeaderMenu';
import ValueHeaderMenu from './valueHeaderMenu';
import MixHeaderMenu from './mixHeaderMenu';
import { columns, mixColumn } from './columnsDef';
import {
  getPlanAnalysisLens,
  selectPlanAnalysis,
  getAllocationGrid,
  removeNodeFromGrid,
} from '../../actions';

export class GridView extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      columnDefs: [...columns],
      selectedPlanAnalysis: undefined,
      selectedAliasSelector: undefined,
    };
  }

  componentDidMount() {
    this.props.getPlanAnalysisLens();
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.tree !== this.props.tree) {
      this.props.getAllocationGrid();
    }
    if (nextProps.mixes !== this.props.mixes) {
      let columnDefs = [];
      _.forEach(nextProps.mixes, mix => {
        mixColumn.headerName = mix;
        mixColumn.field = mix.replace(' ', '').toLowerCase();
        columnDefs.push({ ...mixColumn });
      });
      columnDefs = [...columns, ...columnDefs];
      this.setState({ columnDefs });
    }
  }

  componentDidUpdate() {
    if (this.gridApi) {
      const columns = this.columnApi.getAllColumns();
      if (columns.length > 6) {
        this.gridApi.gridPanel.ensureColumnVisible(columns[6].colId);
      }
    }
  }

  changePlanAnalysisLens = plan => {
    this.props.selectPlanAnalysis(plan);
    this.setState({ selectedPlanAnalysis: plan });
  };

  onGridReady = params => {
    this.gridApi = params.api;
    this.columnApi = params.columnApi;
  };

  onCellEditingStopped = ({ colDef, value, data }) => {
    if (colDef.field.includes('mix') && value === '') {
      this.props.removeNodeFromGrid(data);
    }
  };

  render() {
    return (
      <div className="gridView">
        <Row>
          <Col xs="12">
            <Select
              searchable
              clearable={false}
              value={this.state.selectedPlanAnalysis}
              onChange={this.changePlanAnalysisLens}
              options={this.props.planAnalysisLens}
              className="dropdownPlanAnalysis"
              labelKey="name"
              valueKey="id"
              placeholder="Plan Analysis Lens"
            />
          </Col>
        </Row>
        <Row>
          <Col xs="12">
            <div className="gridContainer ag-fresh">
              <AgGridReact
                columnDefs={this.state.columnDefs}
                rowData={this.props.gridData}
                onGridReady={this.onGridReady}
                onCellEditingStopped={this.onCellEditingStopped}
                headerHeight="35"
                enableSorting
              />
              <NodeNameHeaderMenu />
              <ValueHeaderMenu />
              <MixHeaderMenu />
            </div>
          </Col>
        </Row>
      </div>
    );
  }
}

GridView.propTypes = {
  tree: PropTypes.object.isRequired,
  gridData: PropTypes.array.isRequired,
  mixes: PropTypes.array,
  planAnalysisLens: PropTypes.array.isRequired,
  selectPlanAnalysis: PropTypes.func.isRequired,
  getPlanAnalysisLens: PropTypes.func.isRequired,
  getAllocationGrid: PropTypes.func.isRequired,
  removeNodeFromGrid: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
  return {
    ...state.allocationGrid,
    ...state.allocationTree,
  };
};

export default connect(mapStateToProps, {
  getPlanAnalysisLens,
  selectPlanAnalysis,
  getAllocationGrid,
  removeNodeFromGrid,
})(GridView);
