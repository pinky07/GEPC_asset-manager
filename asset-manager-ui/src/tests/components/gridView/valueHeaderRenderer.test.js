import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { ValueHeaderRenderer } from '../../../components/gridView/valueHeaderRenderer';

let props = {
  column: {
    colDef: {
      headerName: 'Header test',
    },
    addEventListener: jest.fn(),
  },
  gridData: [],
  enableSorting: true,
  progressSort: jest.fn(),
  setSort: jest.fn(),
  displayName: 'test name',
};

describe('valueHeaderRenderer component', () => {
  let wrapper = shallow(<ValueHeaderRenderer {...props} />);

  it('renders correctly', () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  describe('when there is items at gridData', () => {
    beforeEach(() => {
      props.gridData = [1];
      wrapper = shallow(<ValueHeaderRenderer {...props} />);
    });

    it('should renders correctly', () => {
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });
});
