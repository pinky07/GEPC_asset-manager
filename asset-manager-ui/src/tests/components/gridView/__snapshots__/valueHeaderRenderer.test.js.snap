// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`valueHeaderRenderer component renders correctly 1`] = `
<div
  className="ag-cell-label-container    ag-header-cell-sorted-none"
>
  <div
    className="ag-header-cell-label"
    onClick={[Function]}
  >
    <span
      className="ag-header-cell-text"
    >
      Header test
    </span>
    <span
      className="ag-header-icon ag-sort-ascending-icon  ag-hidden"
      onClick={[Function]}
    >
      <span
        className="ag-icon ag-icon-asc"
      />
    </span>
    <span
      className="ag-header-icon ag-sort-descending-icon  ag-hidden"
      onClick={[Function]}
    >
      <span
        className="ag-icon ag-icon-desc"
      />
    </span>
  </div>
</div>
`;

exports[`valueHeaderRenderer component when there is items at gridData should renders correctly 1`] = `
<div
  className="ag-cell-label-container    ag-header-cell-sorted-none"
>
  <div
    className="ag-header-cell-label"
    onClick={[Function]}
  >
    <span
      className="ag-header-cell-text"
    >
      Header test
    </span>
    <span
      className="ag-header-icon ag-sort-ascending-icon  ag-hidden"
      onClick={[Function]}
    >
      <span
        className="ag-icon ag-icon-asc"
      />
    </span>
    <span
      className="ag-header-icon ag-sort-descending-icon  ag-hidden"
      onClick={[Function]}
    >
      <span
        className="ag-icon ag-icon-desc"
      />
    </span>
    <ContextMenuProvider
      className=""
      event="onClick"
      id="value_menu_id"
      renderTag="span"
      style={Object {}}
    >
      <button>
        ...
      </button>
    </ContextMenuProvider>
  </div>
</div>
`;
